packages/monitor-v2/src/monitor-og/MonitorSnapshotProposals.ts

```diff
@@ -0,0 +1,204 @@
+import { Datastore } from "@google-cloud/datastore";
+import assert from "assert";
+import retry, { Options as RetryOptions } from "async-retry";
+import { promises as fsPromises } from "fs";
+import { request } from "graphql-request";
+import { gql } from "graphql-tag";
+
+import { ForkedTenderlyResult, generateForkedSimulation, Logger, MonitoringParams } from "./common";
+import { logSnapshotProposal } from "./MonitorLogger";
+import {
+  GraphqlData,
+  isMatchingSafe,
+  isSnapshotProposalGraphql,
+  SnapshotProposalGraphql,
+  translateToSafeSnap,
+} from "./SnapshotVerification";
+
+let datastoreInstance: Datastore | undefined; // Only used when state is stored in Google Datastore.
+
+// Returns a singleton instance of Google Datastore.
+const getDatastoreInstance = () => {
+  if (!datastoreInstance) {
+    datastoreInstance = new Datastore();
+  }
+  return datastoreInstance;
+};
+
+// Queries Snapshot for all proposals that are still active and have the required plugin.
+// This uses provided retry config, but ultimately returns the error object if the Snapshot query fails after all
+// retries. This also validates returned data and filters only proposals that use either safeSnap or oSnap plugin.
+const getActivePluginProposals = async (
+  plugin: string,
+  url: string,
+  retryOptions: RetryOptions
+): Promise<Array<SnapshotProposalGraphql> | Error> => {
+  const query = gql(/* GraphQL */ `
+    query GetActiveProposals($plugin: String) {
+      proposals(
+        first: 1000 # Maximum number that can be returned by Snapshot (should be enough for active plugin proposals).
+        where: { plugins_contains: $plugin, state: "active" }
+        orderBy: "created"
+        orderDirection: desc
+      ) {
+        id
+        ipfs
+        type
+        choices
+        start
+        end
+        state
+        space {
+          id
+        }
+        scores
+        quorum
+        scores_total
+        plugins
+      }
+    }
+  `);
+
+  // If the GraphQL request fails for any reason, we return an Error object that will be logged by the bot.
+  try {
+    const graphqlData = await retry(
+      () => request<GraphqlData, { plugin: string }>(url, query, { plugin }),
+      retryOptions
+    );
+    // Filter only for proposals that have a properly configured safeSnap or oSnap plugin.
+    return graphqlData.proposals.filter(isSnapshotProposalGraphql);
+  } catch (error) {
+    assert(error instanceof Error, "Unexpected Error type!");
+    return error;
```

Comment:
personally id just not try to catch anything in this call and use promise.allSettled further down to ensure everything gets processed and logged